File: 2024/11/prog_42586_기능개발.py
'''
95 -> <1일 후> 99 -> <2일 후> 100 (이 날 끝에 배포)
93 -> 94 -> 95 -> 96 -> 97 -> 98 -> 99 -> <7일 후> 100 (이 날 끝에 배포)
'''
from collections import deque
import math
def solution(progresses, speeds):
    answer = []
    done_day = deque() # 개발 완료일 
    
    for i in range(len(progresses)):
        # 100 - 93 = 7, 7 / 1 = 7
        # 100 - 30 = 70, 70 / 30 = 2.x -올림-> 3
        p = progresses[i]
        done_day.append(math.ceil((100 - p) / speeds[i]))

    cur = 0
    front = done_day.popleft()
    cnt = 1
    
    while done_day:
        cur = done_day.popleft()

        if cur > front:
            # 이전까지의 기능들 배포 
            answer.append(cnt)
            cnt = 0
            front = cur
            
        cnt += 1
        
    answer.append(cnt)

    return answer

File: 2025/07/prog_42586_기능개발.js
/*
<어떤 기능이 먼저 완성되었더라도 앞에 있는 모든 기능이 완성되지 않으면 배포가 불가능합니다.>

93: 7 남음, 7/1 => 7일 
30: 70 남음, 70/30 => 2+1 => 3일 
55: 45 남음, 45/5 => 9일 
=> [7일, 3일, 9일]

뒤의 것이 현재의 것보다 작거나 같다면, 같이 배포됨 
뒤의 것이 현재의 것보다 크다면, 그보다 더 뒤는 탐색하지 않음 

큐: 오늘 배포할 수 있는 것 
for 
    cur = 현재 남은 일 수 
    큐.push(cur)
    
    if cur > 큐.front
        // 같이 배포 끝 
        answer.push(큐 길이 - 1) 
        큐 초기화 
*/
function solution(progresses, speeds) {
    var answer = [];
    let q = []
    
    for (let i=0; i < progresses.length;i++) {
        const cur = Math.ceil((100-progresses[i]) / speeds[i])
        q.push(cur)
        const front = q[0]
        
        if (cur > front) {
            answer.push(q.length - 1)
            q = [cur]
        }
    }
    
    answer.push(q.length)
    return answer;
}